src/QirRuntime/lib/Tracer/tracer.cpp

```diff
@@ -251,18 +273,76 @@ namespace Quantum
         LayerId layerId = this->TraceSingleQubitOp(id, duration, target);
         return reinterpret_cast<Result>(layerId);
     }
 
+    //------------------------------------------------------------------------------------------------------------------
+    // CTracer::TraceMultiQubitMeasurement
+    //------------------------------------------------------------------------------------------------------------------
     Result CTracer::TraceMultiQubitMeasurement(OpId id, Duration duration, long nTargets, Qubit* targets)
     {
         LayerId layerId = this->TraceMultiQubitOp(id, duration, 0, nullptr, nTargets, targets);
         return reinterpret_cast<Result>(layerId);
     }
 
+    //------------------------------------------------------------------------------------------------------------------
+    // CTracer::FindLatestMeasurementLayer
+    //------------------------------------------------------------------------------------------------------------------
+    LayerId CTracer::FindLatestMeasurementLayer(long count, Result* results)
+    {
+        LayerId latest = INVALID;
+        for (long i = 0; i < count; i++)
+        {
+            const LayerId id = this->GetLayerIdOfSourceMeasurement(results[i]);
+            latest = CTracer::LaterLayer(latest, id);
+        }
+        return latest;
+    }
+
+    //------------------------------------------------------------------------------------------------------------------
+    // CTracer::FenceScope
+    //------------------------------------------------------------------------------------------------------------------
+    CTracer::FenceScope::FenceScope(CTracer* tracer, long count1, Result* rs1, long count2, Result* rs2)
+        : tracer(tracer)
+    {
+        const LayerId fence1 =
+            (rs1 != nullptr && count1 > 0) ? this->tracer->FindLatestMeasurementLayer(count1, rs1) : INVALID;
+        const LayerId fence2 =
+            (rs2 != nullptr && count2 > 0) ? this->tracer->FindLatestMeasurementLayer(count2, rs2) : INVALID;
+
+        this->fence = CTracer::LaterLayer(fence1, fence2);
+        if (this->fence == INVALID)
+        {
+            return;
+        }
+        assert(this->fence < this->tracer->metricsByLayer.size());
+
+        this->tracer->conditionalFences.push_back(this->fence);
+        this->tracer->latestConditionalFence = CTracer::LaterLayer(this->tracer->latestConditionalFence, this->fence);
+    }
+    CTracer::FenceScope::~FenceScope()
+    {
+        if (this->fence == INVALID)
+        {
+            return;
+        }
+
+        vector<LayerId>& fences = this->tracer->conditionalFences;
+        assert(!fences.empty());
+        this->tracer->conditionalFences.pop_back();
+
+        // Update the latest layer (we expect the stack of fences to be shallow so a linear search through it
+        // should be OK).
+        this->tracer->latestConditionalFence =
+            fences.empty() ? INVALID : *std::max_element(fences.begin(), fences.end());

```

Comment:
Question:

The `pop_back()` in line 338 makes me _allow_ that `this->tracer->conditionalFences` _can become empty_ (and the data block is allowed to be deallocated). That is why the `fences` (introduced in line 336) can become (after line 338) a reference to an _empty vector_.
In the line 343 the `std::max_element()` will return (for an empty vector) the `fences.end()` - the iterator pointing to _after_ the data (or even after the deallocated block). The subsequent dereferencing `*` of such an iterator can cause the undefined behavior.

What does guarantee that I'm mistaken?